Support rally landing points

sfuhrer opened the pull request:
This PR is a rebased and slightly modified version of https://github.com/PX4/Firmware/pull/7743 of @bkueng to enable landings at Rally Points during an RTL.

**Describe problem solved by the proposed pull request**
When a RTL is triggered, the vehicle will land at the closest Rally point or home it that's closer. If a mission landing is planned (applies mainly for VTOL or FW), then the DO_LAND_START marker replaces Home, meaning that it will check if that one is closer than any rally point, fly to it and then continue the landing procedure.

**Test data / coverage**
SITL tested.

**Additional context**
This is a first, basic, implementation of Rally points. There is especially some discussion and follow up work needed on when it should land at the closest Rally point, or really come all the way back to home. E.g. if the RTL was triggered by the user or due to RC loss, the expected behavior is that it comes to home. Battery empty RTL's make sense at the closes Rally point. 

Also, if flying a VTOL and a malfunction of the FW system is detected (either automatically or by the user) and it thus is in hover mode, landing at the closes Rally point is crucial as probably the distance to home is too large.


**EDIT**:
I've made some changes:
- nothing changes for RTL_TYPE 0,1,2 (so all that were there before this PR) 
- safe landing points are now only considered if RTL_TYPE=3 ("Return via direct way to whatever is closest: home, mission lading or safe point"
- if DO_LAND_START marker is available, then the distance to this one is compared to home/safe landing points if RTL_TYPE=3

bkueng commented on src/modules/navigator/rtl.cpp:
```diff
@@ -68,6 +69,79 @@ void
+
+		// TODO: handle all possible mission_safe_point.frame cases
+		switch (closest_safe_point.frame) {
+		case 0:
```
I think it's worth commenting which on it is:
```suggestion
		case 0: // MAV_FRAME_GLOBAL
```
Or turning `mission_safe_point_s::frame` into an enum.

bkueng commented on src/modules/navigator/rtl.cpp:
```diff
@@ -68,6 +69,79 @@ void
+			_destination.safe_point_index = closest_index;
+			_destination.lat = closest_safe_point.lat;
+			_destination.lon = closest_safe_point.lon;
+			_destination.alt = closest_safe_point.alt + home_position.alt; // alt of safe point is real to home
```
```suggestion
			_destination.alt = closest_safe_point.alt + home_position.alt; // alt of safe point is rel to home
```

bkueng commented on src/modules/navigator/rtl.cpp:
```diff
@@ -68,6 +69,79 @@ void
+			break;
+
+		default:
+			mavlink_log_critical(_navigator->get_mavlink_log_pub(), "RTL: Rally Point has wrong MAV_FRAME. Landing at HOME.");
```
This is more than 50 chars.
Maybe:
```suggestion
			mavlink_log_critical(_navigator->get_mavlink_log_pub(), "RTL: unsupported MAV_FRAME. Landing at HOME.");
```

julianoes commented:
Jenkins is unhappy:
```
../../src/modules/navigator/rtl.cpp: In member function 'virtual void RTL::on_activation()':
../../src/modules/navigator/rtl.cpp:132:45: error: 'closest_safe_point.mission_safe_point_s::lat' may be used uninitialized in this function [-Werror=maybe-uninitialized]
    _destination.lat = closest_safe_point.lat;

```

hamishwillee commented:
Assuming this goes in, is there any further work to be done to make it work with QGC - ie will upload magically work for PX4 in the same way as for ArduPilot? [Note, I think the upload _probably_ already works, it's just that the points aren't used]

And just to confirm, the handling is currently always to RTL to closest "safe point" (rally point or home)?

sfuhrer commented:
@bkueng I added a new commit to allow for mission landings if the RTL chooses home to land at. Would be cool if you could review that as well.

bkueng commented on src/modules/navigator/rtl_params.c:
```diff
@@ -113,6 +113,8 @@ PARAM_DEFINE_FLOAT(RTL_MIN_DIST, 5.0f);
  * @value 0 Return home via direct path
  * @value 1 Return to a planned mission landing, if available, via direct path, else return to home via direct path
  * @value 2 Return to a planned mission landing, if available, using the mission path, else return to home via the reverse mission path
+ * @value 3 Return via direct path to closest safe point (rally point), if available, else return to home via direct path
```
I would merge this into option 0, because:
- A user who wants safe points would need to change the parameter, adding one more configuration hurdle.
- For a user who does not want safe points it still works, since they would simply not configure any safe points.

sfuhrer replied:
@dagar you mentioned last week during the devcall that you'd prefer the Rally Points not enabled with the RTL_TYPE set to 0 (default). I kind of prefer what beat proposes, to enable landing at Rally Points as soon as there are some planned. 

Actually the consequent way would then be to also remove the RTL_TYPE 4 I added, as this then would be the same as option 1 (It would land at home via mission landing, but at the rally points without).

Of course we could also bring it in with all 5 RTL_TYPEs to have the whole Rally Point feature a bit more hidden, and then change to the original 3 RTL_TYPES when we have different RTL actions depending on what triggered it (eg not go to safe landing for RC loss).

bkueng commented on src/modules/navigator/navigator_main.cpp:
```diff
@@ -561,6 +561,23 @@ Navigator::run()
 
 					break;
 
+				case RTL::RTL_SAFEPOINT_LAND:
```
Since the logic is the same as `RTL::RTL_LAND`, can you merge them?

bkueng commented on src/modules/navigator/rtl.cpp:
```diff
@@ -68,17 +69,97 @@ void
+	}
+
+	// Only chose closest safe point if RTL_TYPE is set to enable safe points, otherwise chose home
+	bool always_return_to_home = !(rtl_type() == RTL_SAFEPOINT_HOME || rtl_type() == RTL_SAFEPOINT_LAND);
```
```suggestion
	bool always_return_to_home = rtl_type() != RTL_SAFEPOINT_HOME && rtl_type() != RTL_SAFEPOINT_LAND;
```

hamishwillee commented:
> -4: RTL_SAFEPOINT_LAND vehicle returns to closest safe point or home, and if it's home
then it uses the planned mission landing (like if RTL_TYPE=1)

What does this mean? A planned mission landing is by definition wherever the mission says it should be, not home.

sfuhrer commented:
> > -4: RTL_SAFEPOINT_LAND vehicle returns to closest safe point or home, and if it's home
> > then it uses the planned mission landing (like if RTL_TYPE=1)
> 
> What does this mean? A planned mission landing is by definition wherever the mission says it should be, not home.
> 
> Also, this appears to be for MC. What happens for FW and VTOL for these settings?

Actually the PR description was outdated, I changed it now. With the current implementation the settings wouldn't change (RTL_TYPE =0,1,2), but it would automatically go to a safe landing point if that one is closer than home. 
If RTL_TYPE=1 is selected, it's currently implemented that still the home position distance is compared to the safe landing points, and if home is closer, the planned mission landing is used. This is of course not optimal as you pointed out @hamishwillee , so I'll try to fix it tomorrow.

bkueng commented on src/modules/navigator/mission.cpp:
```diff
@@ -376,29 +376,45 @@ Mission::set_execution_mode(const uint8_t mode)
+
+			_land_start_available = true;
+			_land_start_index = i;
+			// the DO_LAND_START marker containst no position sp, so take them from the previous mission item
```
```suggestion
			// the DO_LAND_START marker contains no position sp, so take them from the previous mission item
```

bkueng replied:
Is it guaranted that the previous item always contains a position?

sfuhrer replied:
can't say for sure, but it would be a very strange mission if not

bkueng replied:
Can you add a check then? I want to be sure we have well-defined behavior and not random misinterpretations if the assumption does not hold. IMO it's fine to not set `_land_start_available` if the assumption does not hold.

bkueng commented on src/modules/navigator/mission.cpp:
```diff
@@ -376,29 +376,45 @@ Mission::set_execution_mode(const uint8_t mode)
-		    ((missionitem.nav_cmd == NAV_CMD_VTOL_LAND) && _navigator->get_vstatus()->is_vtol) ||
-		    (missionitem.nav_cmd == NAV_CMD_LAND)) {
+		// first check for DO_LAND_START marker
+		if ((missionitem.nav_cmd == NAV_CMD_DO_LAND_START)) {
```
```suggestion
		if (missionitem.nav_cmd == NAV_CMD_DO_LAND_START) {
```

bkueng commented on src/modules/navigator/mission.h:
```diff
@@ -251,6 +254,9 @@ class Mission : public MissionBlock, public ModuleParams
 	// track location of planned mission landing
 	bool	_land_start_available{false};
 	uint16_t _land_start_index{UINT16_MAX};		/**< index of DO_LAND_START, INVALID_DO_LAND_START if no planned landing */
+	float _landing_lat{0.0f};
```
lat/lon should be stored as double:
```suggestion
	double _landing_lat{0.0f};
```

bkueng commented on src/modules/navigator/rtl.h:
```diff
@@ -94,6 +107,27 @@ class RTL : public MissionBlock, public ModuleParams
+		}
+	};
+
+	// RTLType rtl_type{0};
```
Commented code

bkueng commented on src/modules/navigator/rtl.h:
```diff
@@ -94,6 +107,27 @@ class RTL : public MissionBlock, public ModuleParams
+	};
+
+	// RTLType rtl_type{0};
+	RTLPosition _destination{0}; ///< the RTL position to fly to (typically the home position or a safe point)
```
```suggestion
	RTLPosition _destination{}; ///< the RTL position to fly to (typically the home position or a safe point)
```

bkueng commented on src/modules/navigator/rtl.h:
```diff
@@ -94,6 +107,27 @@ class RTL : public MissionBlock, public ModuleParams
+
+	// RTLType rtl_type{0};
+	RTLPosition _destination{0}; ///< the RTL position to fly to (typically the home position or a safe point)
+	RTLDestinationType _destination_type{};
```
Can you explicitly initialize to one of the enums?

bkueng commented on src/modules/navigator/navigator_main.cpp:
```diff
@@ -563,6 +563,22 @@ Navigator::run()
+					}
+
+					// if RTL is set to use a mission landing and mission has a planned landing, then use MISSION to fly there directly
+					if (on_mission_landing() && !get_land_detected()->landed) {
```
Shouldn't we let RTL handle this case as well?

sfuhrer replied:
no don't think so, as in this case we're already in the RTL using mission landing (it set's navigation_mode_new to rtl in the first pass, but to _mission when the RTL using mission landing has startted)

bkueng replied:
Ok, looking at `on_mission_landing()`, that looks right then. Can you merge the code with `case RTL::RTL_LAND:`? It's the same up to the message.

sfuhrer replied:
done

bkueng commented on src/modules/navigator/rtl.cpp:
```diff
@@ -56,6 +57,147 @@ RTL::on_inactive()
+		}
+
+	default: {
+			mavlink_log_critical(_navigator->get_mavlink_log_pub(), "RTL: unsupported RTL_TYPE. Please change RTL_TYPE parameter.");
```
This message is too long (50 chars is the max)

bkueng commented on src/modules/navigator/rtl.cpp:
```diff
@@ -68,17 +210,33 @@ void
+	// output the correct message, depending on where the RTL destination is
+	switch (_destination_type) {
+	case RTL_DESTINATION_HOME:
+		mavlink_and_console_log_info(_navigator->get_mavlink_log_pub(), "RTL: landing at HOME.");
```
Maybe:
```suggestion
		mavlink_and_console_log_info(_navigator->get_mavlink_log_pub(), "RTL: landing at home position.");
```

bkueng commented on src/modules/navigator/rtl.cpp:
```diff
@@ -56,6 +57,147 @@ RTL::on_inactive()
+
+
+	case RTL_CLOSEST: { // choose closest possible landing point (consider home, mission landing and safe points)
+			_destination.set(
```
Consider moving this block into a separate method to reduce the level of indentation.

hamishwillee commented:
The current behaviour for return landing is all in terms of home position. If you're using a rally point are the return altitude and other parameters in the cone relative to the target rally point.

sfuhrer commented:
@bkueng thanks for your review, I've addressed your comments. 

@bresch could you please check that the RTL cone landing feature still works as expected?

hamishwillee commented:
@sfuhrer @bresch Re

> could you please check that the RTL cone landing feature still works as expected?

as above, and please let me know what "as expected" means for all the parameters.

bkueng commented on src/modules/navigator/rtl.cpp:
```diff
@@ -56,6 +57,156 @@ RTL::on_inactive()
+		}
+
+
+	case RTL_CLOSEST: { // choose closest possible landing point (consider home, mission landing and safe points)
```
You don't need the additional brackets (only if you declare a variable inside the block):
```suggestion
	case RTL_CLOSEST: // choose closest possible landing point (consider home, mission landing and safe points)
```

bkueng commented on src/modules/navigator/rtl.cpp:
```diff
@@ -56,6 +57,156 @@ RTL::on_inactive()
+		}
+	}
+
+	// copare to safe landing positions
```
```suggestion
	// compare to safe landing positions
```

bkueng commented on src/modules/navigator/rtl.cpp:
```diff
@@ -56,6 +57,156 @@ RTL::on_inactive()
+			_destination_type = RTL_DESTINATION_MISSION_LANDING;
+
+		} else {
+			_destination_type = RTL_DESTINATION_HOME;
```
Can you set this before `if (_navigator->get_mission_start_land_available()) {`?

bkueng commented on src/modules/navigator/mission.h:
```diff
@@ -251,6 +254,9 @@ class Mission : public MissionBlock, public ModuleParams
 	// track location of planned mission landing
 	bool	_land_start_available{false};
 	uint16_t _land_start_index{UINT16_MAX};		/**< index of DO_LAND_START, INVALID_DO_LAND_START if no planned landing */
+	double _landing_lat{0.0f};
```
```suggestion
	double _landing_lat{0.0};
```

bkueng commented on src/modules/navigator/rtl.h:
```diff
@@ -94,6 +108,26 @@ class RTL : public MissionBlock, public ModuleParams
+	};
+
+	RTLPosition _destination{}; ///< the RTL position to fly to (typically the home position or a safe point)
+	RTLDestinationType _destination_type{RTL_DESTINATION_HOME};
```
Can you move the type into the `RTLPosition` struct?

bkueng commented on src/modules/navigator/mission.h:
```diff
@@ -91,6 +91,9 @@ class Mission : public MissionBlock, public ModuleParams
 	bool get_mission_finished() const { return _mission_type == MISSION_TYPE_NONE; }
 	bool get_mission_changed() const { return _mission_changed ; }
 	bool get_mission_waypoints_changed() const { return _mission_waypoints_changed ; }
+	float get_landing_lat() { return _landing_lat; }
```
```suggestion
	double get_landing_lat() { return _landing_lat; }
```
And in `src/modules/navigator/navigator.h` as well.

LorenzMeier commented:
The review comment commits should not be in upstream later - make sure to squash all commits on merging or do an interactive rebase.

Commits should represent blocks of functionality, not the history of what you did.

mrivi commented:
Flight tested the PR. Behaved as expected. Here's a log https://review.px4.io/plot_app?log=ccd0d34d-f426-444d-ac9f-55526c1be1a9

hamishwillee commented:
@mrivi Can you answer the questions https://github.com/PX4/Firmware/pull/12696#issuecomment-542635962 and https://github.com/PX4/Firmware/pull/12696#issuecomment-543380420 - specifically, which parameters are still relative to home and which are relevant to the rally point. I would assume this would work best if all the parameters are relative to the rally point, including its stated altitude.

hamishwillee commented:
Also, do these apply to all vehicles, or just MC?

sfuhrer commented:
> Also, do these apply to all vehicles, or just MC?

Yes, also applies to VTOL and FW. With RTL_TYPE=3 (aka the new "find closest" landing point in RTL), it will chose whether home, land (resp. the DO_LAND_START waypoint) or a safe landing point is closest. If it's the DO_LAND_START one it will do the mission landing starting at this point, otherwise a "normal" RTL-landing (with that I mean transitioning to MC above landing point for VTOL and descend vertically, and loiter for a FW). Let me know if that makes it clear or if you need some further help (I guess you want to update the user guide?)

sfuhrer commented:
> The current behaviour for return landing is all in terms of home position. If you're using a rally point, are the return altitude and other parameters in the cone relative to the target rally point.?

The return altitude is relative to whatever altitude homer or the safe lading point is set to (resp. to the altitude of the wp at the DO_LAND_START marker). 
@mrivi did you check the cone feature for RTL? It should make no difference if on landing at home/mission landing/rally point.

sfuhrer commented:
> @sfuhrer @bresch Re
> 
> > could you please check that the RTL cone landing feature still works as expected?
> 
> as per [#12696 (comment)](https://github.com/PX4/Firmware/pull/12696#issuecomment-542635962), please let me know what "as expected" means for all the parameters.

So the cone RTL is not yet described anywhere? From what I understand by looking at the implementation is that the user can set the RTL cone angle. The vehicle will then return on the altitude set by the cone, which is a function of horizontal distance to landing point (the alt will though at least be RTL_DESCEND_ALT). @bresch @MaEtUgR please correct me if I'm wrong.

hamishwillee commented:
Thanks @sfuhrer . Yes, that is exactly the clarification I needed - all altitudes are relative to the destination, not home. 

Docs for the cone are here: http://docs.px4.io/master/en/flight_modes/return.html#return_altitude

hamishwillee commented:
I updated RTL param docs in https://github.com/PX4/Firmware/pull/13302. Please review.
